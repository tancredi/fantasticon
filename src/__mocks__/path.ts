const _path = jest.requireActual('path');
const _relative = _path.relative;

const projectDir = _path.resolve(__dirname, '../../');

const resolve = (...paths: string[]) => {
  let path = '';

  for (const cur of paths) {
    path = !path ? normalize(cur) : _path.join(path, normalize(cur));
  }

  if (path.startsWith(projectDir)) {
    path = _path.join('/root/project', path.substr(projectDir.length));
  } else if (path.startsWith('/') && !path.startsWith('/root')) {
    path = _path.join('/root/', path.substr(1));
  } else if (path.startsWith('./')) {
    path = _path.join('/root/project/', path.slice(2));
  } else if (!path.startsWith('/')) {
    path = _path.join('/root/project/', path);
  }

  return normalize(path);
};

const relative = (a: string, b: string) =>
  normalize(_relative(_path.normalize(a), _path.normalize(b)));

const join = (...segments: string[]): string => {
  const trimmed: string[] = [];

  for (let [i, current] of segments.entries()) {
    const isFirst = i === 0;
    const isLast = i === segments.length - 1;

    if (!isFirst && current.startsWith('/')) {
      current = current.substr(1);
    }

    if (!isLast && current.endsWith('/')) {
      current = current.substring(0, current.length - 1);
    }

    trimmed.push(current);
  }

  return trimmed.join('/');
};

const normalize = (path: string) => path.replace(/\\/g, '/');

const isAbsolute = (path: string) => path.startsWith('/root');

module.exports = { resolve, relative, join, isAbsolute, normalize };
