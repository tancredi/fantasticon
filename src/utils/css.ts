import { FontGeneratorOptions } from '../types/generator';
import { FontAssetType } from '../types/misc';
import { getHash } from './hash';

interface RenderSrcOptions {
  formatValue: string;
  getSuffix?: (name: string) => string;
}

const renderSrcOptions: { [key in FontAssetType]: RenderSrcOptions } = {
  [FontAssetType.EOT]: {
    formatValue: 'embedded-opentype',
    getSuffix: () => '#iefix'
  },
  [FontAssetType.WOFF2]: { formatValue: 'woff2' },
  [FontAssetType.WOFF]: { formatValue: 'woff' },
  [FontAssetType.TTF]: { formatValue: 'truetype' },
  [FontAssetType.SVG]: { formatValue: 'svg', getSuffix: name => `#${name}` }
};

export const renderSrcAttribute = (
  { name, fontTypes, fontsUrl }: FontGeneratorOptions,
  font: string | Buffer
) =>
  fontTypes
    .map(fontType => {
      const { formatValue, getSuffix } = renderSrcOptions[fontType];
      const hash = getHash(font.toString('utf8'));
      const suffix = getSuffix ? getSuffix(name) : '';
      return `url("${
        fontsUrl || '.'
      }/${name}.${fontType}?${hash}${suffix}") format("${formatValue}")`;
    })
    .join(',\n');
