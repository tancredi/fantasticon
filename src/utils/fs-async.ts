import { promisify } from 'util';
import * as fs from 'fs';

export const readFile = promisify(fs.readFile);
export const writeFile = promisify(fs.writeFile);
export const stat = promisify(fs.stat);

export const checkPath = async (
  filepath: string,
  type?: 'directory' | 'file'
) => {
  try {
    const result = await stat(filepath);

    if (type) {
      return type === 'directory' ? result.isDirectory() : result.isFile();
    }

    return true;
  } catch {
    return false;
  }
};
