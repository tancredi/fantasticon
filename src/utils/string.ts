export const pluralize = (word: string, amount: number) =>
  amount === 1 ? word : `${word}s`;
