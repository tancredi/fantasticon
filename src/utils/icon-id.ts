import slug from 'slugify';
import { GetIconIdFn } from '../types/misc';
import { removeExtension } from './path';

export const getIconId: GetIconIdFn = ({ relativeFilePath }) =>
  slug(removeExtension(relativeFilePath).replace(/([./\\])+/g, '-'), {
    replacement: '-',
    remove: /['"`]/g
  });
