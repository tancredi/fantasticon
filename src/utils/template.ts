import Handlebars from 'handlebars';
import { readFile } from 'fs/promises';
import { resolve, isAbsolute } from 'path';
import { getHexCodepoint } from './codepoints';

export const TEMPLATE_HELPERS: Record<string, Function> = {
  codepoint: getHexCodepoint
};

export type CompileOptions = {
  helpers?: { [key: string]: (...args: any[]) => string };
};

export const renderTemplate = async (
  templatePath: string,
  context: object,
  options: CompileOptions = {}
) => {
  const absoluteTemplatePath = isAbsolute(templatePath)
    ? templatePath
    : resolve(process.cwd(), templatePath);
  const template = await readFile(absoluteTemplatePath, 'utf8');

  return Handlebars.compile(template)(context, {
    ...options,
    helpers: { ...TEMPLATE_HELPERS, ...(options.helpers || {}) }
  });
};
