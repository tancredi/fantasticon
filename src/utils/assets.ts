import glob from 'glob';
import { promisify } from 'util';
import { resolve, relative, join } from 'path';
import { removeExtension, splitSegments } from '../utils/path';
import { writeFile } from './fs-async';
import { RunnerOptions } from '../types/runner';
import { GeneratedAssets } from '../generators/generate-assets';

export type WriteResult = { content: string | Buffer; writePath: string };

export type WriteResults = WriteResult[];

export interface IconAsset {
  id: string;
  absolutePath: string;
  relativePath: string;
}

export interface AssetsMap {
  [key: string]: IconAsset;
}

export const ASSETS_EXTENSION = 'svg';

export const loadPaths = async (dir: string): Promise<string[]> => {
  const globPath = join(dir, `**/*.${ASSETS_EXTENSION}`).replace(/\\/g,'/');

  const files = await promisify(glob)(globPath, {});

  if (!files.length) {
    throw new Error(`No SVGs found in ${dir}`);
  }

  return files;
};

const failForConflictingId = (
  { relativePath: pathA, id }: IconAsset,
  { relativePath: pathB }: IconAsset
): void => {
  throw new Error(
    `Conflicting result from 'getIconId': '${id}' - conflicting input files:\n` +
      [pathA, pathB].map(fpath => `  - ${fpath}`).join('\n')
  );
};

export const loadAssets = async ({
  inputDir,
  getIconId
}: RunnerOptions): Promise<AssetsMap> => {
  const paths = await loadPaths(inputDir);
  const out = {};
  let index = 0;

  for (const path of paths) {
    const relativePath = relative(resolve(inputDir), resolve(path));
    const parts = splitSegments(relativePath);
    const basename = removeExtension(parts.pop());
    const absolutePath = resolve(path);
    const iconId = getIconId({
      basename,
      relativeDirPath: join(...parts),
      absoluteFilePath: absolutePath,
      relativeFilePath: relativePath,
      index
    });

    const result: IconAsset = { id: iconId, relativePath, absolutePath };

    if (out[iconId]) {
      failForConflictingId(out[iconId], result);
    }

    out[iconId] = result;

    index++;
  }

  return out;
};

export const writeAssets = async (
  assets: GeneratedAssets,
  { name, pathOptions = {}, outputDir }: RunnerOptions
) => {
  const results: WriteResults = [];

  for (const ext of Object.keys(assets)) {
    const filename = [name, ext].join('.');
    const writePath = pathOptions[ext] || join(outputDir, filename);
    results.push({ content: assets[ext], writePath });
    await writeFile(writePath, assets[ext]);
  }

  return results;
};
