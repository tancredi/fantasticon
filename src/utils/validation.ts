import { checkPath } from './fs-async';
import { CodepointsMap } from './codepoints';

export const parseNumeric = (value: string) => {
  const out = Number(value);

  if (
    (typeof value !== 'number' && typeof value !== 'string') ||
    Number.isNaN(out)
  ) {
    throw new Error(`${value} is not a valid number`);
  }

  return out;
};

export const parseString = (value: string) => {
  if (typeof value !== 'string') {
    throw new Error(`${value} is not a string`);
  }

  return value;
};

export const parseFunction = (value: Function) => {
  if (typeof value !== 'function') {
    throw new Error(`${value} is not a function`);
  }

  return value;
};

export const listMembersParser = <T extends string>(allowedValues: T[]) => (
  values: string[]
) => {
  for (const value of values) {
    if (!allowedValues.includes(value as any)) {
      throw new Error(
        [
          `${value} is not valid`,
          `accepted values are: ${allowedValues.join(', ')}`
        ].join(' - ')
      );
    }
  }

  return values as T[];
};

export const parseCodepoints = (codepoints: Object): CodepointsMap => {
  for (const key of Object.keys(codepoints)) {
    if (typeof codepoints[key] === 'string') {
      codepoints[key] = parseInt(codepoints[key], 16);
    }
  }
  return codepoints as CodepointsMap;
};

export const removeUndefined = (object: Object) => {
  for (const key of Object.keys(object)) {
    if (typeof object[key] === 'undefined') {
      delete object[key];
    }
  }

  return object;
};

export const parseBoolean = (val: any) => {
  if (typeof val === 'string' && ['1', '0', 'true', 'false'].includes(val)) {
    return val === 'true' || val === '1';
  } else if (typeof val === 'number' && [0, 1].includes(val)) {
    return val === 1;
  } else if (typeof val === 'boolean') {
    return val;
  }

  throw new Error(`must be a boolean value`);
};

const skipIfMatching = (match: any) => (fn: (value: any, cur?: any) => any) => (
  val: any
) => (val === match ? match : fn(val));

export const parseDir = async (dirname: string) => {
  if ((await checkPath(dirname, 'directory')) === false) {
    throw new Error(`${dirname} is not a directory`);
  }

  return dirname;
};

export const nullable = skipIfMatching(null);

export const optional = skipIfMatching(undefined);
