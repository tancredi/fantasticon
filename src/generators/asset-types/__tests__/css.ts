import { resolve } from 'path';
import cssGen from '../css';
import { renderSrcAttribute } from '../../../utils/css';

const renderSrcMock = renderSrcAttribute as any as jest.Mock;

const mockOptions = {
  name: 'test-font',
  prefix: 'tf',
  tag: 'b',
  codepoints: { 'my-icon': 0xf101 },
  assets: { 'my-icon': null },
  templates: {
    css: resolve(__dirname, '../../../../templates/css.hbs')
  }
} as any;

jest.mock('../../../utils/css', () => ({
  renderSrcAttribute: jest.fn(() => '"::src-attr::"')
}));

describe('`CSS` asset generator', () => {
  beforeEach(() => {
    renderSrcMock.mockClear();
  });

  test('renders CSS correctly with prefix and tag name options', async () => {
    expect(
      await cssGen.generate(mockOptions, Buffer.from(''))
    ).toMatchSnapshot();
  });

  test('renders CSS correctly with `selector` option', async () => {
    expect(
      await cssGen.generate(
        { ...mockOptions, selector: '.my-selector' },
        Buffer.from('')
      )
    ).toMatchSnapshot();
  });

  test('calls renderSrcAttribute correctly and includes its return value in the rendered template', async () => {
    const fontBuffer = Buffer.from('::svg-content::');

    const result = await cssGen.generate(mockOptions, fontBuffer);

    expect(renderSrcMock).toHaveBeenCalledTimes(1);
    expect(renderSrcMock).toHaveBeenCalledWith(mockOptions, fontBuffer);
    expect(result).toContain('::src-attr::');
  });

  test('renders expected selector blocks', async () => {
    const css = await cssGen.generate(mockOptions, Buffer.from(''));

    expect(css).toContain('b[class^="tf-"]:before, b[class*=" tf-"]:before {');
    expect(css).toContain('.tf-my-icon:before {');
  });

  test('renders expected selector blocks with `selector` option', async () => {
    const css = await cssGen.generate(
      { ...mockOptions, selector: '.my-selector' },
      Buffer.from('')
    );

    expect(css).toContain('.my-selector:before {');
    expect(css).toContain('.my-selector.tf-my-icon:before {');
  });
});
