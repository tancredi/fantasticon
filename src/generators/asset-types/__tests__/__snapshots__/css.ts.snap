// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`CSS\` asset generator renders CSS correctly with \`selector\` option 1`] = `
"@font-face {
    font-family: "test-font";
    src: "::src-attr::";
}

.my-selector:before {
    font-family: test-font !important;
    font-style: normal;
    font-weight: normal !important;
    font-variant: normal;
    text-transform: none;
    line-height: 1;
    -webkit-font-smoothing: antialiased;
    -moz-osx-font-smoothing: grayscale;
}

.my-selector.tf-my-icon:before {
    content: "\\f101";
}
"
`;

exports[`\`CSS\` asset generator renders CSS correctly with prefix and tag name options 1`] = `
"@font-face {
    font-family: "test-font";
    src: "::src-attr::";
}

b[class^="tf-"]:before, b[class*=" tf-"]:before {
    font-family: test-font !important;
    font-style: normal;
    font-weight: normal !important;
    font-variant: normal;
    text-transform: none;
    line-height: 1;
    -webkit-font-smoothing: antialiased;
    -moz-osx-font-smoothing: grayscale;
}

.tf-my-icon:before {
    content: "\\f101";
}
"
`;
